hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java

```diff
@@ -0,0 +1,378 @@
+      LoggerFactory.getLogger(SnapshotDeletingService.class);
+
+  // Use only a single thread for Snapshot Deletion. Multiple threads would read
+  // from the same table and can send deletion requests for same snapshot
+  // multiple times.
+  private static final int SNAPSHOT_DELETING_CORE_POOL_SIZE = 1;
+  private final ClientId clientId = ClientId.randomId();
+  private final AtomicLong runCount;
+
+  private final OzoneManager ozoneManager;
+  private final OmSnapshotManager omSnapshotManager;
+  private final SnapshotChainManager chainManager;
+  private final AtomicBoolean suspended;
+  private final OzoneConfiguration conf;
+  private final AtomicLong successRunCount;
+  private final long snapshotDeletionPerTask;
+
+  public SnapshotDeletingService(long interval, long serviceTimeout,
+      OzoneManager ozoneManager) throws IOException {
+    super("SnapshotDeletingService", interval, TimeUnit.MILLISECONDS,
+        SNAPSHOT_DELETING_CORE_POOL_SIZE, serviceTimeout);
+    this.ozoneManager = ozoneManager;
+    this.omSnapshotManager = ozoneManager.getOmSnapshotManager();
+    this.chainManager = ozoneManager.getSnapshotChainManager();
+    this.runCount = new AtomicLong(0);
+    this.successRunCount = new AtomicLong(0);
+    this.suspended = new AtomicBoolean(false);
+    this.conf = ozoneManager.getConfiguration();
+    this.snapshotDeletionPerTask = conf
+        .getLong(SNAPSHOT_DELETING_LIMIT_PER_TASK,
+        SNAPSHOT_DELETING_LIMIT_PER_TASK_DEFAULT);
+  }
+
+  private class SnapshotDeletingTask implements BackgroundTask {
+
+    @Override
+    public BackgroundTaskResult call() throws Exception {
+      if (!shouldRun()) {
+        return BackgroundTaskResult.EmptyTaskResult.newResult();
+      }
+
+      runCount.incrementAndGet();
+
+      Table<String, SnapshotInfo> snapshotInfoTable =
+          ozoneManager.getMetadataManager().getSnapshotInfoTable();
+      try (TableIterator<String, ? extends Table.KeyValue
+          <String, SnapshotInfo>> iterator = snapshotInfoTable.iterator()) {
+        iterator.seekToFirst();
+
+        long snapshotLimit = snapshotDeletionPerTask;
+
+        while (iterator.hasNext() && snapshotLimit > 0) {
+          SnapshotInfo snapInfo = iterator.next().getValue();
+          SnapshotInfo.SnapshotStatus snapshotStatus =
+              snapInfo.getSnapshotStatus();
+
+          // Only Iterate in deleted snapshot
+          if (!snapshotStatus.equals(
+              SnapshotInfo.SnapshotStatus.SNAPSHOT_DELETED)) {
+            continue;
+          }
+
+          OmSnapshot omSnapshot = (OmSnapshot) omSnapshotManager
+              .checkForSnapshot(snapInfo.getVolumeName(),
+                  snapInfo.getBucketName(),
+                  getSnapshotPrefix(snapInfo.getName()));
+
+          Table<String, RepeatedOmKeyInfo> snapshotDeletedTable =
+              omSnapshot.getMetadataManager().getDeletedTable();
+
+          if (snapshotDeletedTable.getEstimatedKeyCount() == 0) {
```

Comment:
I'm not familiar with "getEstimatedKeyCount()".  Since it is "estimated" is there any chance that this will cause us to skip keys that should be gc'd?

Reply:
`snapshotDeletedTable.isEmpty()` should be of better use here, as `snapshotDeletedTable.getEstimatedKeyCount()` can be inaccurate. Updating the PR.